File: lit_cli/main/pull.py
import requests
import toml
import yaml
from pathlib import Path
from halo import Halo
import click

DATA_PATH = Path(".lit_env_data.toml")
ENV_PATH = Path(".env")
CONFIG_PATH = Path("./main/config.yaml")


def load_token():
    if DATA_PATH.exists():
        data = toml.load(DATA_PATH)
        return data.get("auths", {}).get("token")
    return None


def load_project_id():
    if DATA_PATH.exists():
        data = toml.load(DATA_PATH)
        return data.get("active_project", {}).get("id")
    return None


#
#
# def load_api_url():
#     if CONFIG_PATH.exists():
#         with open(CONFIG_PATH, "r") as f:
#             config = yaml.safe_load(f)
#             return config.get("server", {}).get("api_url")
#     return None

API_URL = "http://localhost:8080"


@click.command()
def pull():
    # print("jsks")
    token = load_token()
    project_id = load_project_id()

    if not token or not project_id or not API_URL:
        click.secho("❌ Missing token, project ID, or API URL.", fg="red")
        return

    spinner = Halo(text="Fetching project .env data...", spinner="dots")
    spinner.start()

    try:
        response = requests.get(
            f"{API_URL}/projects/pull-env-data/{project_id}",
            headers={"Authorization": f"Bearer {token}"},
            timeout=30,
        )
        spinner.stop()

        if response.status_code != 200:
            click.secho(
                f"❌ Failed to fetch project: {response.status_code}",
                fg="red",
            )
            click.secho(response.text, fg="red")
            return

        env_content = response.json().get("data", "")

        if not env_content.strip():
            click.secho(
                "⚠️ No environment data found for this project.",
                fg="yellow",
            )
            return

        with open(ENV_PATH, "w") as f:
            f.write(env_content)

        click.secho(
            f"✅ Environment variables written to {ENV_PATH}",
            fg="green",
        )

    except requests.RequestException as e:
        spinner.fail("🚨 Network error.")
        click.secho(str(e), fg="red")

File: lit_cli_v2/handlers/pull.go
package handlers

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"

	"github.com/briandowns/spinner"
	"github.com/spf13/cobra"
)

var PullCmd = &cobra.Command{
	Use:   "pull",
	Short: "Pull environment variables from the server",
	Run: func(cmd *cobra.Command, args []string) {
		pullEnv()
	},
}

func loadToken() string {
	data, err := os.ReadFile(".lit_env_data.toml")
	if err != nil {
		return ""
	}
	type Auths struct {
		Email string `toml:"email"`
		Token string `toml:"token"`
	}
	type Config struct {
		Auths Auths `toml:"auths"`
	}

	// Simple TOML parse (for real use, you'd use BurntSushi/toml or similar)
	return extractTOMLToken(data)
}

func loadProjectID() string {
	data, err := os.ReadFile(".lit_env_data.toml")
	if err != nil {
		return ""
	}
	return extractTOMLProjectID(data)
}

func extractTOMLToken(data []byte) string {
	// NOTE: Replace with real TOML parsing
	if str := string(data); len(str) > 0 {
		start := "token = \""
		end := "\""
		s := findBetween(str, start, end)
		return s
	}
	return ""
}

func extractTOMLProjectID(data []byte) string {
	// NOTE: Replace with real TOML parsing
	if str := string(data); len(str) > 0 {
		start := "id = \""
		end := "\""
		s := findBetween(str, start, end)
		return s
	}
	return ""
}

func findBetween(s, start, end string) string {
	startIdx := len(start) + len(s[:len(s)-len(start)]) - len(s)
	if idx := len(s[:len(s)-len(end)]); idx >= startIdx {
		return s[startIdx:idx]
	}
	return ""
}

func pullEnv() {
	API_URL := "http://localhost:8080"
	token := loadToken()
	projectID := loadProjectID()

	if token == "" || projectID == "" {
		fmt.Println("❌ Missing token or project ID.")
		return
	}

	s := spinner.New(spinner.CharSets[14], 100*time.Millisecond)
	s.Suffix = " Fetching project .env data..."
	s.Start()

	url := fmt.Sprintf("%s/projects/pull-env-data/%s", API_URL, projectID)
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("Authorization", "Bearer "+token)

	client := http.Client{Timeout: 30 * time.Second}
	res, err := client.Do(req)
	if err != nil {
		s.FinalMSG = "🚨 Network error.\n"
		s.Stop()
		fmt.Println(err)
		return
	}
	defer res.Body.Close()

	if res.StatusCode != 200 {
		body, _ := io.ReadAll(res.Body)
		s.FinalMSG = fmt.Sprintf("❌ Failed to fetch project: %s\n", res.Status)
		s.Stop()
		fmt.Println(string(body))
		return
	}

	var result struct {
		Data string `json:"data"`
	}
	if err := json.NewDecoder(res.Body).Decode(&result); err != nil {
		s.FinalMSG = "❌ Failed to parse server response\n"
		s.Stop()
		os.Exit(1)
	}

	s.Stop()

	if result.Data == "" {
		fmt.Println("⚠️ No environment data found for this project.")
		return
	}

	os.WriteFile(".env", []byte(result.Data), 0644)
	fmt.Println("✅ Environment variables written to .env")
}


func init() {
	rootCmd.AddCommand(PullCmd)
}
